From 86d48c24bdf96494a8a017d15c368574794d580a Mon Sep 17 00:00:00 2001
From: Darren Oh <9472-darrenoh@users.noreply.drupalcode.org>
Date: Sun, 5 Jan 2025 12:45:04 -0500
Subject: [PATCH] Issue #3497485: Fix remaining duplication of batch jobs

---
 .../drupal_cms_installer.profile              | 25 +++++++++++++++----
 1 file changed, 20 insertions(+), 5 deletions(-)

diff --git a/project_template/web/profiles/drupal_cms_installer/drupal_cms_installer.profile b/project_template/web/profiles/drupal_cms_installer/drupal_cms_installer.profile
index 1ad09758..e3bf9519 100644
--- a/project_template/web/profiles/drupal_cms_installer/drupal_cms_installer.profile
+++ b/project_template/web/profiles/drupal_cms_installer/drupal_cms_installer.profile
@@ -240,9 +240,10 @@ function _drupal_cms_installer_password_value(&$element, $input, FormStateInterf
 function drupal_cms_installer_apply_recipes(array &$install_state): array {
   // If the installer ran before but failed mid-stream, don't reapply any
   // recipes that were successfully applied.
+  $recipes_applied = \Drupal::state()->get(RecipeAppliedSubscriber::STATE_KEY, []);
   $recipes_to_apply = array_diff(
     $install_state['parameters']['recipes'],
-    \Drupal::state()->get(RecipeAppliedSubscriber::STATE_KEY, []),
+    $recipes_applied,
   );
 
   // If we've already applied all the chosen recipes, there's nothing to do.
@@ -255,17 +256,31 @@ function drupal_cms_installer_apply_recipes(array &$install_state): array {
   $batch = install_profile_modules($install_state);
   $batch['title'] = t('Setting up your site');
 
-  $recipe_operations = [];
+  // If we started applying recipes, profile modules are already installed.
+  if (!empty($recipes_applied)) {
+    $batch['operations'] = [];
+  }
+
+  $recipe_operations = $applied_operations = [];
 
-  foreach ($recipes_to_apply as $name) {
+  foreach ($install_state['parameters']['recipes'] as $name) {
+    $applied = in_array($name, $recipes_applied);
     $recipe = InstalledVersions::getInstallPath('drupal/' . $name);
     $recipe = Recipe::createFromDirectory($recipe);
-    $recipe_operations = array_merge($recipe_operations, RecipeRunner::toBatchOperations($recipe));
+    if ($applied) {
+      $applied_operations = array_merge($applied_operations, RecipeRunner::toBatchOperations($recipe));
+    }
+    else {
+      $recipe_operations = array_merge($recipe_operations, RecipeRunner::toBatchOperations($recipe));
+    }
   }
 
   // Only do each recipe's batch operations once.
   foreach ($recipe_operations as $operation) {
-    if (in_array($operation, $batch['operations'], TRUE)) {
+    if (
+      in_array($operation, $batch['operations'], TRUE) ||
+      in_array($operation, $applied_operations, TRUE)
+    ) {
       continue;
     }
     else {
-- 
GitLab

